core/src/main/java/org/apache/iceberg/io/ReadBuilder.java

```diff
@@ -0,0 +1,123 @@
+   *
+   * @param newStart the start position for this read
+   * @param newLength the length of the range this read should scan
+   */
+  B split(long newStart, long newLength);
+
+  /** Read only the given columns. */
+  B project(Schema newSchema);
+
+  /**
+   * Pushes down the {@link Expression} filter for the reader to prevent reading unnecessary
+   * records. Some readers might not be able to filter some part of the expression. In this case the
+   * reader might return unfiltered or partially filtered rows. It is the caller's responsibility to
+   * apply the filter again.
+   *
+   * @param newFilter the filter to set
+   * @param filterCaseSensitive whether the filtering is case-sensitive or not
+   */
+  default B filter(Expression newFilter, boolean filterCaseSensitive) {
+    // Skip filtering if not available
+    return (B) this;
+  }
+
+  /**
+   * Pushes down the {@link Expression} filter for the reader to prevent reading unnecessary
+   * records. Some readers might not be able to filter some part of the exception. In this case the
+   * reader might return unfiltered or partially filtered rows. It is the caller's responsibility to
+   * apply the filter again. The default implementation sets the filter to be case-sensitive.
+   *
+   * @param newFilter the filter to set
+   */
+  default B filter(Expression newFilter) {
+    return filter(newFilter, true);
+  }
+
+  /**
+   * Sets configuration key/value pairs for the reader. Reader builders should ignore configuration
+   * keys not known for them.
+   */
+  default B set(String key, String value) {
+    // Skip configuration if not applicable
+    return (B) this;
+  }
+
+  /**
+   * Enables reusing the containers returned by the reader. Decreases pressure on GC. Readers could
+   * decide to ignore the user provided setting if is not supported by them.
+   */
+  default B reuseContainers() {
+    // Skip container reuse configuration if not applicable
+    return (B) this;
+  }
+
+  /**
+   * Accessors for constant field values. Used for calculating values in the result which are coming
+   * from metadata, and not coming from the data files themselves. The keys of the map are the
+   * column ids, the values are the accessors generating the values.
+   */
+  B constantFieldAccessors(Map<Integer, ?> constantFieldAccessors);
+
+  /** Sets a mapping from external schema names to Iceberg type IDs. */
+  B withNameMapping(NameMapping newNameMapping);
+
+  /**
+   * Sets the file encryption key used for reading the file. If encryption is not supported by the
+   * reader then an exception should be thrown.
+   */
+  default B withFileEncryptionKey(ByteBuffer encryptionKey) {
+    throw new UnsupportedOperationException("Not supported");
+  }
+
+  /**
+   * Sets the additional authentication data prefix for encryption. If encryption is not supported
+   * by the reader then an exception should be thrown.
+   */
+  default B withAADPrefix(ByteBuffer aadPrefix) {
```

Comment:
inconsistent naming. some methods use `with`, while some don't

Reply:
These are the methods which are not pushed down to the `Appender`. They were different in the previous API, and I have kept it so, but I'm open to change it. This would mean more code change when migrating to the new API, but I am fine if the community decides so

Reply:
Also, shall we use aadPrefix/aADPrefix/withAadPrefix in the new API?

Currently it is `withAADPrefix`

Reply:
Went ahead, and removed the `with` prefix from everywhere.
It's much more consistent, and the users should not know about the internal details.